Automattic/wp-calypso
# Add condition to closeUrl for block editor to handle checklist referrals

### Summary

After being referred to the block editor from the checklist, this PR implements directing back to the checklist when the block editor close button is clicked.

![backtochecklist](https://user-images.githubusercontent.com/8892849/63870508-16fe2380-c988-11e9-866c-09f92e0e7a47.gif)

* Overrides close button URL similarly to how the previous implementation handles FSE Template post type close buttons.
* A bit of refactoring to _kill the lets_.

<hr>

### Outstanding and questions

- [x] "View Pages"/"View Posts" tooltip on the close button is a bit wonky when you're not actually going back to Pages/Posts. This will probably need to be overridden in the iframe server. Which leads me to the question...
- [x] Does it make more sense to just completely override the close button for active checklist tasks? Similarly to FSE. Might create a better call to action when working through the checklist and solve the above problem as well.
<img width="359" alt="Screen Shot 2019-08-28 at 11 49 31 AM" src="https://user-images.githubusercontent.com/8892849/63871473-f33bdd00-c989-11e9-91cb-e9dbbf8b53e7.png">

- [x] Does the default URL for the close button make more sense to just go to the previous route, as opposed to the Posts/Pages listings? (Assuming that a previous route exists, otherwise default to Posts/Pages.)

^ Above questions answered. Outstanding:

- [x] Move functionality into a selector and utilize the 'actionLog' to look further back in the route stack.
- [x] Add a label to back button for checklist navigation and remove "Back to Post/Page" tooltip.
- [x] Ensure it is possible to navigate back to checklist after editing a template part.

<hr>

### Testing

* Spin up a local Calypso testing environment. You will also have to build the wpcom block editor.
* Open a site with an active checklist.
* Test the individual flows:
    * Test the Checklist-to-block editor flow: Ie, by clicking "Try It" under the "Update your homepage" task in the checklist. Expect that the close/back button in the block editor takes you back to the Checklist.
    * Test the neutral back button flow: Ie, by selecting a post from the Posts page or going directly to the block editor by pasting a URL in the address bar. Expect that the close/black button in the block editor takes you back to the Posts (or Pages) listing page.
    * Test the pre-existing FSE flow: Ie, by going to a page, then selecting "Edit Header". Try the back button from the Template part. Expect that it brings you back to the parent page editor.

Fixes #35452
